el-patch-template.el

```diff
@@ -0,0 +1,554 @@
+                                          (list new-match))
+                                  next-step-fn
+                                  table literal))))
+          (el-patch-template--process (car form) (car template)
+                                      nil consp-next-step table
+                                      literal))))
+     ((and (vectorp template) (vectorp form))
+      (el-patch-template--process (append form nil) ;; convert to list
+                                  (append template nil)
+                                  nil
+                                  (lambda (new-match remainder-form)
+                                    (when remainder-form
+                                      ;; Must be complete match
+                                      (throw 'no-match nil))
+                                    (funcall next-step-fn
+                                             (if match
+                                                 (append match
+                                                         (list (apply
+                                                                'vector
+                                                                new-match)))
+                                               (apply 'vector new-match))
+                                             remainder-form))
+                                  table
+                                  literal))
+     ((null template) ;; nothing else to match
+      (funcall next-step-fn match form))
+     ((or (member template '(...)) (equal template form))
+      ;; A Complete match.
+      (funcall next-step-fn (append match form) nil))
+     (t
+      (or (when-let ((symbol (gethash template table))
+                     (symbol-next-step
+                      (lambda (new-match remainder-form)
+                        ;; First put the match in the table.
+                        (let ((old-entry (gethash template table)))
+                          (puthash template
+                                   (cons
+                                    (car symbol)
+                                    new-match)
+                                   table)
+                          ;; Then process the next step, adding the
+                          ;; template to the match
+                          (condition-case _
+                              (funcall next-step-fn
+                                       (if match
+                                           (cons match template)
+                                         template)
+                                       remainder-form)
+                            ('no-match
+                             ;; Ultimately, the matching did not
+                             ;; work, so undo the symbol resolution
+                             (puthash template old-entry table)
+                             ;; and rethrow
+                             (throw 'no-match nil)))))))
```

Comment:
The processing functions `el-patch-template--process-el-patch` and `el-patch-template--process`  accept a `next-step-fn` function handle which is called with the matched and unmatched forms. This is important for `el-patch-*` directives because at this point the matched forms are processed to be a valid `el-patch-*` directive. This works well, but the functions has these lambda definitions which are sometimes long and might make the function hard to read. 

Do you think these definitions are fine? Or should I write functions like `el-patch-template--process-next-step/let` which return these lambda definitions given some parameters? This would also remove unnecessary capturing of parameters, if any.

Reply:
Yeah, I think this is perfectly good. Of course we can always refactor later if it turns out to be necessary, but I think it looks great as is! I wouldn't worry about lexical capturing---that should have negligible impact, even if it feels unclean.